src/hsm_adapter.cpp

```diff
@@ -0,0 +1,79 @@
+#include "hsm_adapter.h"
+
+#include <cstring>
+#include <stdexcept>
+#include <dlfcn.h>
+
+CryptokiAdapter::CryptokiAdapter(CryptokiConfig config) : _config(config) {
+
+    // Load the PKCS#11 library dynamically
+    void* _pkcs11Lib = dlopen(_config.libraryFile.c_str(), RTLD_NOW);
+    if (!_pkcs11Lib) {
+        std::string errMsg = "Error loading PKCS#11 library: " + std::string(dlerror());
+        throw std::runtime_error(errMsg);
+    }
+
+    // Get a pointer to the PKCS#11 function list
+    CK_RV (*C_GetFunctionList)(CK_FUNCTION_LIST_PTR_PTR) =
+        (CK_RV(*)(CK_FUNCTION_LIST_PTR_PTR))dlsym(_pkcs11Lib, "C_GetFunctionList");
+    if (!C_GetFunctionList) {
+        dlclose(_pkcs11Lib);
+        std::string errMsg = "Error getting function pointer: " + std::string(dlerror());
+        throw std::runtime_error(errMsg);
+    }
+
+    CK_RV rv = C_GetFunctionList(&_pFunctionList);
+    if (rv != CKR_OK) {
+        dlclose(_pkcs11Lib);
+        std::string errMsg = "C_GetFunctionList Error: " + std::to_string(rv);
+        throw std::runtime_error(errMsg);
+    }
+
+}
+
+CryptokiAdapter::~CryptokiAdapter() {
+    dlclose(_pkcs11Lib);
+}
+
+void CryptokiAdapter::injectSeedRandom(std::vector<uint8_t> random) {
+
+    // Initialize the library
+    CK_RV rv = _pFunctionList->C_Initialize(NULL);
+    if (rv != CKR_OK) {
+        std::string errMsg = "C_Initialize Error: " + std::to_string(rv);
+        throw std::runtime_error(errMsg);
+    }
+
+    // Open a session
+    CK_SESSION_HANDLE hSession;
+    rv = _pFunctionList->C_OpenSession(_config.slotID, CKF_SERIAL_SESSION, NULL, NULL, &hSession);
```

Comment:
Andrew - could you walk through why we want to initialize/open session/login for every injection instead of doing that once at the beginning?

Reply:
I prefer not to keep the session open and logged in for a long time. I'm not sure if it's important to keep them short lived, but I think it's a reasonable place to start. I'm not worried about performance overhead for this application at this time. We still need to work with customers on requirements.